server/client/js/main.js

```diff
@@ -6,64 +6,158 @@ let users = db.collection('users');
-
-            // Set RTDB online key
-            let presenceRef = rtdb.ref(`/status/${userID}`)
-            presenceRef.set('online').then(() => {
-                console.log("added user presence ref");
-            }).catch((error) => {
-                console.log("error adding user presence ref:", error);
-            });
-            
-            // On RTDB set offline for deletion from firestore
-            $(window).on('beforeunload', (e) => {
-                presenceRef.set('offline')
-                    .then(() => console.log("set user offline"))
-                    .catch((error) => console.error("unable to set user offline:", error));
-                e.returnValue = "";
-                return "";
+        console.log("user signed in", user);
+        $("#signIn").hide();
+        $("#signUp").hide();
+        if (user.displayName) {
+            $("#user").text(`${user.displayName} (${user.email})`);
+            $("#content").show();
+        }
+
+        // Retain reference to user doc in firestore
+        userDoc = users.doc(user.uid);
+
+        // Get user doc info
+        userDoc.get()
+            .then(docSnapshot => {
+                let data = docSnapshot.data();
+                if (data) {
+                    lastOperation = data.lastOperation;
+                }
+            })
+            .catch(e => {
+                console.log(`unable to get user doc info: ${e.message}`);
             });
 
+        if (!socket) {
             // Connect to server via websocket
             socket = new Socket(`${location.host}/ws`);
 
             // Set up callbacks
-            socket.register("operationUpdate", (m) => {
+            socket.register("operationUpdate", m => {
                 $("#operations").append(`<p>${JSON.stringify(m.operation)}</p>`);
             });
-            socket.register("clearState", (m) => {
+            socket.register("clearState", m => {
                 $("#operations").empty();
             });
-            socket.register("numMembersUpdate", (m) => {
+            socket.register("numMembersUpdate", m => {
                 $("#numMembers").text(m.numMembers);
             });
 
-            // Announce this user 
+            // Announce this user
             socket.addEventListener("open", () => {
-                socket.send({
+                socket.sendWithResponse({
                     "id": uuidv4(),
                     "type": "announce",
-                    "userID": userID
-                });
+                    "userID": user.uid
+                })
+                    .then(res => {
+                        if(res.error) {
+                            // Client already connected
+                            alert("You're already logged in in another tab.");
+                            $("#content").remove();
+                        }
+                    })
+                    .catch(error => console.log("unable to announce:", error));
```

Comment:
The `announce` type message now has a response which returns an `error` property in the body should the user be signed-in and have the page open in another tab. I suggest we don't support this, but we could figure that out.

Reply:
Do you mean we shouldnt support limiting users to one tab? Is it just that it adds complexity?

Reply:
I suggest we _don't_ allow users to operate in more than one tab. Not sure how we would resolve conflicts and it doesn't make much sense to me.

Reply:
Yeah I agree, that would just be them circumventing the rules + making it complicated for us